target/linux/ramips/patches-5.4/0302-mt7621-nand-add-dma-support.patch

```diff
@@ -0,0 +1,1347 @@
++
++static int mt7621_nfc_read_oob_raw(struct nand_chip *nand, int page)
++{
++	return mt7621_read_oob_raw(nand, nand->oob_poi, page, 0);
++}
+ 
++/* write page without check */
++static int mt7621_write_page_hwecc(struct nand_chip *nand, const uint8_t *buf,
++				   const uint8_t *oob, int page)
++{
++	struct mt7621_nfc *nfc = nand_get_controller_data(nand);
++	struct mtd_info *mtd = nand_to_mtd(nand);
++	dma_addr_t addr;
++	size_t len = mtd->writesize;
++	int ret;
++
++	/* prepare dma */
++	addr = dma_map_single(nfc->dev, (void *)buf, len, DMA_TO_DEVICE);
++	ret = dma_mapping_error(nfc->dev, addr);
++	if (ret) {
++		dev_warn(nfc->dev, "dma mapping error\n");
++		return -EINVAL;
++	}
++	nfi_write32(nfc, NFI_STRADDR, addr);
++	init_completion(&nfc->nfi_done);
++	init_completion(&nfc->ecc_done);
++
++	nand_prog_page_begin_op(nand, page, 0, NULL, 0);
++
++	/* init nfi/ecc */
+ 	nfi_write16(nfc, NFI_CNFG, (CNFG_OP_CUSTOM << CNFG_OP_MODE_S) |
+-		    CNFG_READ_MODE);
++		   CNFG_AUTO_FMT_EN | CNFG_HW_ECC_EN | CNFG_DMA_MODE);
++	mt7621_ecc_encoder_op(nfc, true);
+ 
++	mt7621_nfc_write_fdm(nfc, oob);
++
++	/* start nfi transfer */
+ 	nfi_write16(nfc, NFI_CON,
+-		    CON_NFI_BRD | (nand->ecc.steps << CON_NFI_SEC_S));
++		    CON_NFI_BWR | (nand->ecc.steps << CON_NFI_SEC_S));
++	nfi_write16(nfc, NFI_STRDATA, STR_DATA);
+ 
+-	for (i = 0; i < nand->ecc.steps; i++) {
+-		/* Read data */
+-		if (buf)
+-			mt7621_nfc_read_data(nfc, page_data_ptr(nand, buf, i),
+-					     nand->ecc.size);
+-		else
+-			mt7621_nfc_read_data_discard(nfc, nand->ecc.size);
+-
+-		/* Read FDM */
+-		mt7621_nfc_read_data(nfc, oob_fdm_ptr(nand, i), NFI_FDM_SIZE);
+-
+-		/* Read ECC parity data */
+-		mt7621_nfc_read_data(nfc, oob_ecc_ptr(nfc, i),
+-				     nfc->spare_per_sector - NFI_FDM_SIZE);
++	/* wait nfi interrupt */
++	ret = wait_for_completion_timeout(&nfc->nfi_done,
++					  msecs_to_jiffies(500));
++	if (!ret) {
++		dev_warn(nfc->dev, "write ahb/dma done timeout\n");
++		ret = -ETIMEDOUT;
++		goto err;
++	}
++
++	/* wait ecc interrupt */
++	ret = wait_for_completion_timeout(&nfc->ecc_done,
++					  msecs_to_jiffies(500));
++	if (!ret) {
++		dev_warn(nfc->dev, "decode timeout\n");
++		ret = -ETIMEDOUT;
++		goto err;
+ 	}
+ 
++	ret = 0;
++err:
++	mt7621_ecc_encoder_op(nfc, false);
++	dma_unmap_single(nfc->dev, addr, len, DMA_TO_DEVICE);
+ 	nfi_write16(nfc, NFI_CON, 0);
+ 
+-	return 0;
+-}
++	if (ret)
++		return ret;
+ 
+-static int mt7621_nfc_read_oob_hwecc(struct nand_chip *nand, int page)
+-{
+-	return mt7621_nfc_read_page_hwecc(nand, NULL, 1, page);
+-}
++	ret = nand_prog_page_end_op(nand);
++	if ((ret == 0) && oob)
++		ret = mt7621_write_oob_raw(nand, oob, page, 1);
+ 
+-static int mt7621_nfc_read_oob_raw(struct nand_chip *nand, int page)
+-{
+-	return mt7621_nfc_read_page_raw(nand, NULL, 1, page);
++	return ret;
+ }
+ 
+-static int mt7621_nfc_check_empty_page(struct nand_chip *nand, const u8 *buf)
++static int mt7621_block_page_erase(struct nand_chip *nand, int page)
```

Comment:
the mtk ecc not only protect page data. also protect some oob data(mtk call it fdm).
when you use jffs2 format flash. jffs2 will set cleanmarker on oob only. and that oob is ecc protected. it means oob have cleanmarker and ecc data. so when you only write page data. it will regenerate ecc data part cause it damage.
so when this case happen. need to backup block data. erase block. write page data and cleanmarker together. to let the ecc data right.
normal driver not protect oob data. so don't have this problem.

Reply:
using jffs2 on nand is a bad idea. It doesn't do a good job of wear-levelling and doesn't have a great recovery mechanism. (with page-cache enabled, a power cut during write operation usually results in a broken fs)
I think we don't need such a workaround for jffs2 in openwrt since we don't use it anyway.

Reply:
there is another case. write page first then update oob(fdm) later. will cause the same problem. but i don't know if this case can happen. if sure don't need this workaround. i can remove it.